Add TransferImageBuffer message

mattwoodrow opened the pull request:
#### 08a58d6e99250f434fab7ccbe8f89d3602ddcc7e
<pre>
Add TransferImageBuffer message
<a href="https://bugs.webkit.org/show_bug.cgi?id=249111">https://bugs.webkit.org/show_bug.cgi?id=249111</a>
&lt;rdar://problem/103232081&gt;

Reviewed by Kimmo Kinnunen.

This implements transfering the backend of a RemoteImageBuffer from its existing RemoteRenderingBackend/WorkQueue to its new RRB.
This will likely be replaced in the future with the new way of sharing resources in the GPUP.

When we create a new ImageBuffer in the WebProcess, and want to transfer contents, we send the message across the connection for the
new buffer.
The GPUP side dispatches messages to the main thread to find the RemoteRenderingBackend, dispatches to the existing WorkQueue to take
ownership of the old buffer and dispatches the result back to the new WorkQueue to complete the transfer.

RemoteImageBuffer sends the same didCreateImageBufferBackend message back once the buffer is transferred (as happens when a remote buffer
is allocated), so that WebProcess attempts to block and wait for the buffer to be available succeed.

* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::transferToNewContext):
(WebCore::ImageBufferBackend::parameters):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::transferToNewContext):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::didClose):
(WebKit::GPUConnectionToWebProcess::findRemoteRenderingBackend):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::~RemoteImageBuffer):
(WebKit::RemoteImageBuffer::create):
(WebKit::RemoteImageBuffer::createTransfer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::create): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::transferImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
(WebKit::RemoteRenderingBackend::resourceOwner const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheImageBuffer):
(WebKit::RemoteResourceCache::cachedImageBuffer const):
(WebKit::RemoteResourceCache::takeImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::sendToStreamWithAsyncReply):

Canonical link: <a href="https://commits.webkit.org/257769@main">https://commits.webkit.org/257769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f3edb56a62ae4e81408be01b858205ec913d595

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99914 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109267 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169505 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103916 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9967 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86380 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92365 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107170 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105681 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34256 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22200 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2887 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23713 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2843 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46085 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5331 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8976 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43187 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4695 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
```diff
@@ -24,6 +24,7 @@
 
 messages -> RemoteRenderingBackend NotRefCounted Stream {
     CreateImageBuffer(WebCore::FloatSize logicalSize, WebCore::RenderingMode renderingMode, WebCore::RenderingPurpose renderingPurpose, float resolutionScale, WebCore::DestinationColorSpace colorSpace, enum:uint8_t WebCore::PixelFormat pixelFormat, WebCore::RenderingResourceIdentifier renderingResourceIdentifier)
+    TransferImageBuffer(WebKit::RenderingBackendIdentifier existingBackendIdentifier, WebCore::RenderingResourceIdentifier existingIdentifier, WebCore::RenderingResourceIdentifier newIdentifier) -> (bool success) NotStreamEncodable
```
I don't think this needs to be not stream encodable.

kkinnunen-apple commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -608,6 +608,14 @@ void GPUConnectionToWebProcess::releaseGraphicsContextGLForTesting(GraphicsConte
 }
 #endif
 
+RemoteRenderingBackend* GPUConnectionToWebProcess::findRemoteRenderingBackend(RenderingBackendIdentifier renderingBackendIdentifier)
```
I'd probably call this just `remoteRenderingBackend`.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -175,11 +175,36 @@ void RemoteRenderingBackend::createImageBuffer(const FloatSize& logicalSize, Ren
+
+        existingBackend->dispatch([existingBackend = Ref { *existingBackend }, existingQualified, newBackend, newQualified, completionHandler = WTFMove(completionHandler)] () mutable {
+            if (auto existingBuffer = existingBackend->m_remoteResourceCache.takeImageBuffer(existingQualified)) {
+                newBackend->dispatch([existingBuffer = existingBuffer.releaseNonNull(), newBackend, newQualified, completionHandler = WTFMove(completionHandler)] () mutable {
```
I don't see anything that would be enforcing that this dispatch happens before the newIdentifier is used in the other backend. This means that likely this function `transferImageBuffer` cannot be async, it must be sync.

kkinnunen-apple reviewed (approved):
I'm marking this up as r+ to let you progress if you so wish. Maybe if you double-check the logic regarding whether
the transfer function should be synchronous or asynchronous. 

I think overall this patch makes the ImageBuffer related classes go in wrong direction, a bit more convoluted and
potentially more incorrect direction. If you have time, I'd be happy to see more thought on how to reconcile the 
problems listed below, perhaps by fixing the underlying problems in the ImageBitmap abstraction that uses ImageBuffer,
so that not so much RRB -related logic leaks into ImageBuffer logic.

For the patch:
- I think I mentioned this, already, but: I think this patch should have been done in such a way that all the functions be called 
  in the intended points, and no dead code would have been added. The idea is that even though the calls are in correct position,
  they could have been implemented in non-optimal way via downloading the image to native image, transferring the native 
  image cross threads if needed, and drawing to the pixel buffer back. This way it would be easier to assess what was the end 
  product. Now it's hard to see even with the full patchset, and splitting the patches actually makes it harder to review.
  The idea was to split the patches in self-contained, working parts. Currently the main payload of this patch, for example,
  doesn't seem to be invoked at all and it never will be used, rather further modified before being called ?
  As it stands currently, we cannot ask for "what would be the test to test this added code" to see that it works, as the code is
  not even used.

- I don't immediately see why transfer involves taking the backing of RemoteImageBuffer and destroying the RemoteImageBuffer, then
  recreating it. I don't immediately see why it couldn't be moved from one cache to another.

- The TransferImageBuffer() -> (bool success)
  I do not immediately see how the flag for success can be used correctly. If the notification comes, say, 1ms, 100ms or 10s or 10mins after calling,
  I don't see how the context can be valid in such a way that it can be acted upon.
  I think better method for the success is such that it is handled the same way as creation success / failure should be handled.

For the patchset:

- I don't immediately see why SerializedImageBuffer should be ImageBuffer. These do not share any methods together,
  none of the ImageBuffer interfaces can be available for SerializedImageBuffer. I'm thinking that it may be
  that the ImageBitmapBacking is now used as a wrong abstraction.
  Maybe ImageBitmap should not have a ImageBitmapBacking, seems like a redundant class. Instead the values of backing
  should just be part of ImageBitmap.
  Maybe ImageBitmapBacking should represent the serialized ImageBuffer, e.g. it should be a virtual base that has a method
  that takes in the proper interfaces that are able to construct normal in-process ImageBuffer for non-GPUP case or
  instantiate the correct remote object with the connection context that the interface implementation holds.

- I don't immediately see why it's necessary or ok to make ImageBuffer a RefCounted instead of current ThreadSafeRefCounted.

- Currently the ImageBuffer interface is already convoluted and the WebCore::ImageBuffer part contains a lot of stuff that
  are about the remote rendering, which should not be part of the interface at all. This patchset adds a lot more, which
  I don't think is good situation. This is the main concern I have with the work, regardless of whether it happens to work
  currently or not.



I would maybe implement the transfer as follows:
```
messages -> RemoteRenderingBackend NotRefCounted Stream {
     // ...
     MoveToSerializedBuffer(WebCore::RenderingResourceIdentifier sourceImageBuffer, ImageBufferSerializationWriteReference destination)
     MoveToImageBuffer(ImageBufferSerializationWriteReference source, WebCore::RenderingResourceIdentifier destinationImageBuffer)

}

// MoveToImageBufferSerialization adopts the source and creates the destination

// MoveToImageBuffer adopts the source and creates the destination



class RemoteImageBufferSerializationProxy {
private:
     ImageBufferSerializationReferenceTracker m_referenceTracker;
};

// Note: not a ImageBuffer
class RemoteImageBufferSerialization {
     ... 
private:
     Ref<RemoteImageBuffer> m_object;
};

class RemoteImageBufferSerializationObjectHeap : public ThreadSafeRefCounted<RemoteImageBufferSerializationObjectHeap> {
     ...
private:
     ThreadSafeObjectHeap<RemoteImageBufferSerialization, Ref<RemoteImageBuffer>> m_objects;
};

class GPUConnectionToWebProcess {
     ..
     auto remoteImageBufferSerializations() { return m_remoteImageBufferSerializations; }
     RemoteImageBufferSerializationObjectHeap m_heap;
};
```

E.g the transfer function wouldn't inject the object ad-hoc to the other remote rendering backend, as there is
nobody holding that ref in the proxy backend.

Instead, the ref should transfer from image buffer to the transfer object, and then the destination RRBP/RRB should adopt
the transfer object and transfer the ref from the transfer object to the destination object.

mattwoodrow committed:
Add TransferImageBuffer message
https://bugs.webkit.org/show_bug.cgi?id=249111
<rdar://problem/103232081>

Reviewed by Kimmo Kinnunen.

This implements transfering the backend of a RemoteImageBuffer from its existing RemoteRenderingBackend/WorkQueue to its new RRB.
This will likely be replaced in the future with the new way of sharing resources in the GPUP.

When we create a new ImageBuffer in the WebProcess, and want to transfer contents, we send the message across the connection for the
new buffer.
The GPUP side dispatches messages to the main thread to find the RemoteRenderingBackend, dispatches to the existing WorkQueue to take
ownership of the old buffer and dispatches the result back to the new WorkQueue to complete the transfer.

RemoteImageBuffer sends the same didCreateImageBufferBackend message back once the buffer is transferred (as happens when a remote buffer
is allocated), so that WebProcess attempts to block and wait for the buffer to be available succeed.

* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::transferToNewContext):
(WebCore::ImageBufferBackend::parameters):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::transferToNewContext):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::didClose):
(WebKit::GPUConnectionToWebProcess::findRemoteRenderingBackend):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::~RemoteImageBuffer):
(WebKit::RemoteImageBuffer::create):
(WebKit::RemoteImageBuffer::createTransfer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::create): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::transferImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
(WebKit::RemoteRenderingBackend::resourceOwner const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheImageBuffer):
(WebKit::RemoteResourceCache::cachedImageBuffer const):
(WebKit::RemoteResourceCache::takeImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::sendToStreamWithAsyncReply):

Canonical link: https://commits.webkit.org/257769@main
